cluster-autoscaler/core/scaledown/actuation/actuator_test.go

```diff
@@ -937,6 +961,170 @@ func TestStartDeletion(t *testing.T) {
+				sync.Mutex
+				requests map[string]bool
+			}{requests: test.failedRequests}
+			gotFailedRequest := func(nodeGroupId string) bool {
+				failedRequest.Lock()
+				defer failedRequest.Unlock()
+				val, _ := failedRequest.requests[nodeGroupId]
+				return val
+			}
+			deletedResult := make(chan string)
+			fakeClient := &fake.Clientset{}
+			provider := testprovider.NewTestCloudProvider(nil, func(nodeGroupId string, node string) error {
+				if gotFailedRequest(nodeGroupId) {
+					return fmt.Errorf("SIMULATED ERROR: won't remove node")
+				}
+				deletedResult <- nodeGroupId
+				return nil
+			})
+			// 2d array represent the waves of pushing nodes to delete.
+			deleteNodes := [][]*apiv1.Node{}
+			for ng, numNodes := range test.numNodesToDelete {
+				provider.InsertNodeGroup(ng)
+				ng.SetCloudProvider(provider)
+				for i, num := range numNodes {
+					nodes := generateNodes(num, ng.Id())
+					if len(deleteNodes) <= i {
+						deleteNodes = append(deleteNodes, nodes)
+					} else {
+						deleteNodes[i] = append(deleteNodes[i], nodes...)
+					}
+					for _, node := range nodes {
+						provider.AddNode(ng.Id(), node)
+					}
+				}
+			}
+			opts := config.AutoscalingOptions{
+				MaxScaleDownParallelism:        10,
+				MaxDrainParallelism:            5,
+				MaxPodEvictionTime:             0,
+				DaemonSetEvictionForEmptyNodes: true,
+			}
+			registry := kube_util.NewListerRegistry(nil, nil, nil, nil, nil, nil, nil, nil, nil, nil)
+			ctx, err := NewScaleTestAutoscalingContext(opts, fakeClient, registry, provider, nil, nil)
+			if err != nil {
+				t.Fatalf("Couldn't set up autoscaling context: %v", err)
+			}
+			csr := clusterstate.NewClusterStateRegistry(provider, clusterstate.ClusterStateRegistryConfig{}, ctx.LogRecorder, NewBackoff())
+			ndt := deletiontracker.NewNodeDeletionTracker(0)
+			actuator := Actuator{
+				ctx: &ctx, clusterState: csr, nodeDeletionTracker: ndt,
+				nodeDeletionBatcher: NewNodeBatcherDeleter(&ctx, csr, ndt, deleteInterval),
+				evictor:             Evictor{EvictionRetryTime: 0, DsEvictionRetryTime: 0, DsEvictionEmptyNodeTimeout: 0, PodEvictionHeadroom: DefaultPodEvictionHeadroom},
+			}
+			for _, nodes := range deleteNodes {
+				actuator.StartDeletion(nodes, []*apiv1.Node{}, time.Now())
+				time.Sleep(deleteInterval)
+			}
+			wantDeletedNodes := 0
+			for _, num := range test.wantSuccessfulDeletion {
+				wantDeletedNodes += num
+			}
+			gotDeletedNodes := map[string]int{
+				"test-ng-1": 0,
+				"test-ng-2": 0,
+				"test-ng-3": 0,
+			}
+			for i := 0; i < wantDeletedNodes; i++ {
+				select {
+				case ngId := <-deletedResult:
+					if value, ok := gotDeletedNodes[ngId]; ok {
+						gotDeletedNodes[ngId] = value + 1
+					} else {
+						gotDeletedNodes[ngId] = 1
+					}
+				case <-time.After(1 * time.Second):
+					t.Errorf("Timeout while waiting for deleted nodes.")
+					break
+				}
+			}
+			if diff := cmp.Diff(test.wantSuccessfulDeletion, gotDeletedNodes); diff != "" {
```

Comment:
What does this test verify that the test above doesn't? I'd imagine we want to test if the nodes actually get batched, but this test doesn't verify that at all, since the cloud provider hook processes nodes one-by-one. I'd try to assert if all nodes from one wave get deleted in the same API call. That will probably require some fiddling around with the cloud provider, but IMO it's needed.

Reply:
The waves is show the set of nodes for which one we call actuator.StartDeletion, but only nodes that belong to one nodeGroup are deleted in one API call. 
The previous test is testing batcher with 0 seconds interval and this test is testing batcher with >0 seconds interval. So this test is testing the e2e deletion. The tests in Batches are aimed to test the correctness of batching behaviour.